pandaserver/dataservice/finisher.py

```diff
@@ -0,0 +1,193 @@
+            job_dict = {"subFiles": [file_dict]}
+            json_dict["files"]["output"].append(job_dict)
+        return json.dumps(json_dict)
+
+    def update_job_output_report(self, job, failed_files: List[str], no_out_files: List[str]):
+        """
+        This function updates the job output report.
+
+        Parameters:
+        job (JobSpec): The job specification.
+        failed_files (list): List of failed files.
+        no_out_files (list): List of files with no output.
+        """
+        json_data = self.create_json_doc(job, failed_files, no_out_files)
+        record_status = "finished" if not failed_files else "failed"
+        tmp_ret = self.task_buffer.updateJobOutputReport(
+            panda_id=job.PandaID,
+            attempt_nr=job.attemptNr,
+            data=json_data,
+        )
+        if not tmp_ret:
+            self.task_buffer.insertJobOutputReport(
+                panda_id=job.PandaID,
+                prod_source_label=job.prodSourceLabel,
+                job_status=record_status,
+                attempt_nr=job.attemptNr,
+                data=json_data,
+            )
+
+    def check_file_status(self, job):
+        """
+        This function checks the status of the files for the job.
+
+        Parameters:
+        job (JobSpec): The job specification.
+
+        Returns:
+        tuple: A tuple containing three elements:
+            - bool: True if all files are ready, False otherwise.
+            - list: A list of failed files.
+            - list: A list of files with no output.
+        """
+        tmp_log = LogWrapper(_logger,
+                             f"check_file_status-{datetime.datetime.now(datetime.timezone.utc).replace(tzinfo=None).isoformat('/')}")
+        failed_files = []
+        no_out_files = []
+        for file in job.Files:
+            if file.type in ('output', 'log'):
+                if file.status == "failed":
+                    failed_files.append(file.lfn)
+                elif file.status == "nooutput":
+                    no_out_files.append(file.lfn)
+                elif file.status != "ready":
+                    tmp_log.debug(f"Job: {job.PandaID} file:{file.lfn} {file.status} != ready")
+                    return False, failed_files, no_out_files
+        return True, failed_files, no_out_files
+
+    # main
+    def run(self):
+        """
+        Starts the thread to finish transferring jobs
+        """
+        tmp_log = LogWrapper(_logger,
+                             f"run-{datetime.datetime.now(datetime.timezone.utc).replace(tzinfo=None).isoformat('/')}")
+        # start
+        try:
+            by_call_back = False
+            tmp_log.debug(f"start: {self.job.PandaID}")
+            # update input files
+            panda_ids = [self.job.PandaID]
+            tmp_log.debug(f"IDs: {panda_ids}")
+            if len(panda_ids) != 0:
+                # get job
+                if self.job is None:
```

Comment:
Maybe I'm getting this wrong, but the big block you deleted in our last discussion was checking "if self.job is None:". If that block was never being executed, the small "if self.job is None:" can also be removed. I might be wrong, please think it through

Reply:
You are right, I am rechecking